src/lib/ip.rs

```diff
@@ -81,141 +113,254 @@ impl IpConf {
+            )
+            .await?;
+        }
+    }
+
+    Ok(())
+}
+
+async fn apply_ip_conf(
+    handle: &rtnetlink::Handle,
+    nl_addr_msgs: Option<&HashMap<String, AddressMessage>>,
+    iface_index: u32,
+    ip_conf: Option<&IpConf>,
+    cur_ip_conf: Option<IpConf>,
+    ip_family: IpFamily,
+) -> Result<(), NisporError> {
+    // TODO: Can we use single queue?
+    match (ip_conf, cur_ip_conf) {
+        (None, None) => (),
+        (None, Some(_)) => {
+            // Desire would like to remove all address except IPv6 link local
+            // address
+            if let Some(nl_addr_msgs) = nl_addr_msgs {
+                for (address_full, nl_addr_msg) in nl_addr_msgs.iter() {
+                    match ip_family {
+                        IpFamily::Ipv4 => {
+                            if !is_ipv6_addr(&address_full) {
+                                handle
+                                    .address()
+                                    .del(nl_addr_msg.clone())
+                                    .execute()
+                                    .await?;
+                            }
+                        }
+                        IpFamily::Ipv6 => {
+                            if is_ipv6_addr(&address_full)
+                                && !is_ipv6_unicast_link_local(&address_full)
+                            {
+                                handle
+                                    .address()
+                                    .del(nl_addr_msg.clone())
+                                    .execute()
+                                    .await?;
+                            }
+                        }
+                    };
+                }
+            }
+        }
+        (Some(ip_conf), None) => {
+            // Desire would like to add more address
+            for addr_conf in &ip_conf.addresses {
+                handle
+                    .address()
+                    .add(
+                        iface_index,
+                        ip_addr_str_to_enum(&addr_conf.address)?,
+                        addr_conf.prefix_len,
+                    )
+                    .execute()
+                    .await?;
+            }
+        }
+        (Some(ip_conf), Some(cur_ip_conf)) => {
+            let mut cur_ip_addr_confs = HashSet::new();
+            let mut des_ip_addr_confs = HashSet::new();
+            for des_addr in &ip_conf.addresses {
+                des_ip_addr_confs.insert(IpAddrConf {
+                    address: des_addr.address.clone(),
+                    prefix_len: des_addr.prefix_len,
+                });
+            }
+            for cur_addr in &cur_ip_conf.addresses {
+                cur_ip_addr_confs.insert(IpAddrConf {
+                    address: cur_addr.address.clone(),
+                    prefix_len: cur_addr.prefix_len,
+                });
+            }
+            let has_ipv6_link_local_in_desire = if ip_family == IpFamily::Ipv4 {
```

Comment:
This does not have to be `ip_family == IpFamily::Ipv4`

Reply:
It save some cpu time on iteration of `is_ipv6_unicast_link_local_full` over ipv4 addresses.